src/navx_node.cpp

```diff
@@ -67,7 +77,62 @@ int main(int argc, char** argv) {
   ros::spin();
 }
 
+void calculate_covariance(boost::array<double, 9> &orientation_mat, boost::array<double, 9> &ang_vel_mat, boost::array<double, 9> &accel_mat) {
+  int count = std::min(seq-1, covar_samples);
+  if (count < 2) {
+    return;
+  }
+  Pose avg = {};
+  // Calculate averages
+  for (int i = 0; i < count; i++) {
+    for (int j = 0; j < 3; j++) {
+      avg.ypr[j] += poseHistory[i].ypr[j];
+      avg.ang_vel[j] += poseHistory[i].ang_vel[j];
+      avg.accel[j] += poseHistory[i].accel[j];
+    }
+  }
+  for (int j = 0; j < 3; j++) {
+    avg.ypr[j] /= count;
+    avg.ang_vel[j] /= count;
+    avg.accel[j] /= count;
+  }
+  // Calculate covariance
+  for (int x = 0; x < 3; x++) {
+    for (int y = 0; y < 3; y++) {
+      int idx = 3*x + y;
+      orientation_mat[idx] = 0;
+      ang_vel_mat[idx] = 0;
+      accel_mat[idx] = 0;
+      // Average mean error difference
+      for (int i = 0; i < count; i++) {
+        orientation_mat[idx] += (poseHistory[i].ypr[x] - avg.ypr[x]) * (poseHistory[i].ypr[y] - avg.ypr[y]);
+        ang_vel_mat[idx] += (poseHistory[i].ang_vel[x] - avg.ang_vel[x]) * (poseHistory[i].ang_vel[y] - avg.ang_vel[y]);
+        accel_mat[idx] += (poseHistory[i].accel[x] - avg.accel[x]) * (poseHistory[i].accel[y] - avg.accel[y]);
+      }
+      // Normalize
+      orientation_mat[idx] /= count - 1;
+      ang_vel_mat[idx] /= count - 1;
+      accel_mat[idx] /= count - 1;
+    }
+  }
+
```

Comment:
If you used an external source for algo/reference, would be good to include links, either here or at end of file.